### ISSUE TITLE
Implicit conversion loses integer precision 'sqlite3_int64' (aka 'long long') to 'long'

### ISSUE THREAD

**Issue by Kemenger** [ccgus/fmdb#586]
In Xcode 8.3.2 the FMDatabase file method "- (long)valueLong:(void *)value" tips "Implicit conversion loses integer precision 'sqlite3_int64' (aka 'long long') to 'long'"

**Comment by martinflorek** [comment]
How did you solve this? I am getting same error when building v2.7 via Carthage:

`src/fmdb/FMDatabase.m:1495:12: error: implicit conversion loses integer precision: 'sqlite3_int64' (aka 'long long') to 'long' [-Werror,-Wshorten-64-to-32]`

**Comment by robertmryan** [comment]
@Kemenger - I'm trying to reproduce this issue, but you closed this before I got a chance to take a look at it. Does this mean that you found a resolution? I must confess that I'm not seeing this warning. What device were you doing this on? What version of Xcode? Swift or Objective-C app? I'd like to be able to reproduce the warning in order to address it.

@ccgus - Setting aside the fact that I'm having trouble reproducing the warning, there is an inconsistency here. For example, `FMResultSet` method `longForColumnIndex` appears to just silence this warning by casting the result of `sqlite3_column_int64` to `long` before returning it. (That means that if you run it on 32-bit platform, the 64-bit result from `sqlite3_column_int64` is cast to 32-bit result. But it also has a `longLongForColumnIndex` method that returns, obviously, a `long long`.

I'm seeing a couple of approaches:

1. I guess I could do the same for the new `valueLong` as we did for `longForColumnIndex`, e.g. something like:

    <img width="882" alt="screenshot 2017-06-02 11 36 45" src="https://cloud.githubusercontent.com/assets/1530768/26739713/c96f708a-4787-11e7-8ad6-7766a74e4235.png">

    I'm not crazy about perpetuating three permutations, `int`, `long`, and `long long` when SQLite really only has two native types, `int` and `int_64`. This is especially confusing from Swift. For example, look at how Swift renders the existing `FMResultSet` methods `intForColumn`, `longForColumn`, and `longLongIntForColumn`:

    ```swift
    let rs = try! db.executeQuery("select bar from foo", values: nil)
    while rs.next() {
        let a = rs.int(forColumn: "bar")          // returns Swift Int32
        let b = rs.long(forColumn: "bar")         // returns Swift Int
        let c = rs.longLongInt(forColumn: "bar")  // returns Swift Int64
    }
    ```

2. An alternative is to just make it return `long long`:

    <img width="883" alt="screenshot 2017-06-02 11 42 58" src="https://cloud.githubusercontent.com/assets/1530768/26740008/ccecb208-4788-11e7-983a-696a6f08c7ab.png">

    And in Swift, that's rendered as `Int64`. And we'd lose a permutation that returns a Swift `Int` type. Also, it's not consistent with `FMResultSet`'s `longForColumnIndex` and `longLongForColumnIndex`. We might want to do the same change over there if we did this.

There are other possible approaches, but I think I'll stop there.

While I'm not crazy about it, I'm inclined to adopt the first approach because it's at least consistent. Do you have thoughts?

**Comment by ccgus** [comment]
I'm ok with valueLong switching to (long long). That bit of code is only ever used by folks who write custom functions, and presumably that's advanced enough that the developer will know what's going on.

I'm also OK with the first option, it would be consistent as you pointed out.

I can't get the warning to happen either though.

**Comment by limouren** [comment]
I am getting this error in a Swift project with Carthage. My Xcode version is 8.3.2 (8E2002).

**Comment by limouren** [comment]
Here is an example project to reproduce this issue:

* [fmdb27compile.zip](https://github.com/ccgus/fmdb/files/1049481/fmdb27compile.zip)

My bash log reproducing the error:

```sh
➜  fmdb27compile carthage update
*** Fetching fmdb
*** Checking out fmdb at "v2.7"
*** xcodebuild output can be found in /var/folders/hq/cgb7547x1l38kx3sn0rb2rjr0000gn/T/carthage-xcodebuild.V3hYmy.log
*** Building scheme "FMDB MacOS" in fmdb.xcodeproj
*** Building scheme "FMDB iOS" in fmdb.xcodeproj
Build Failed
	Task failed with exit code 65:
	/usr/bin/xcrun xcodebuild -project /Users/MASKED_USER/fmdb27compile/Carthage/Checkouts/fmdb/fmdb.xcodeproj -scheme FMDB\ iOS -configuration Release -derivedDataPath /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7 -sdk iphoneos ONLY_ACTIVE_ARCH=NO BITCODE_GENERATION_MODE=bitcode CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES clean build

This usually indicates that project itself failed to compile. Please check the xcodebuild log for more details: /var/folders/hq/cgb7547x1l38kx3sn0rb2rjr0000gn/T/carthage-xcodebuild.V3hYmy.log
```

The mentioned xcodebuild log file: [carthage-xcodebuild.V3hYmy.log.zip](https://github.com/ccgus/fmdb/files/1049482/carthage-xcodebuild.V3hYmy.log.zip)

The lines that I believed to be relevant (L189-197):

```sh
CompileC /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/Objects-normal/armv7/FMDatabase.o src/fmdb/FMDatabase.m normal armv7 objective-c com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/MASKED_USER/fmdb27compile/Carthage/Checkouts/fmdb
    export LANG=en_US.US-ASCII
    export PATH="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/Users/MASKED_USER/.cargo/bin:/Users/MASKED_USER/google-cloud-sdk/bin:/Users/MASKED_USER/.rbenv/shims:/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/MASKED_USER/go/bin:/Users/MASKED_USER/.composer/vendor/bin/:/Users/MASKED_USER/.local/bin"
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x objective-c -arch armv7 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -std=gnu99 -fobjc-arc -fmodules -gmodules -fmodules-cache-path=/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/ModuleCache -fmodules-prune-interval=86400 -fmodules-prune-after=345600 -fbuild-session-file=/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/ModuleCache/Session.modulevalidation -fmodules-validate-once-per-build-session -Wnon-modular-include-in-framework-module -Werror=non-modular-include-in-framework-module -Xclang -fmodule-implementation-of -Xclang FMDB -Wno-trigraphs -fpascal-strings -Os -fno-common -Werror -Wno-missing-field-initializers -Wno-missing-prototypes -Werror=return-type -Wunreachable-code -Wno-implicit-atomic-properties -Werror=deprecated-objc-isa-usage -Werror=objc-root-class -Wno-arc-repeated-use-of-weak -Wduplicate-method-match -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wconditional-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -Wno-selector -Wno-strict-selector-match -Wundeclared-selector -Wno-deprecated-implementations -DNS_BLOCK_ASSERTIONS=1 -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS10.3.sdk -fstrict-aliasing -Wprotocol -Wdeprecated-declarations -miphoneos-version-min=9.2 -g -Wno-sign-conversion -Winfinite-recursion -fembed-bitcode -iquote /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/FMDB-generated-files.hmap -I/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/FMDB-own-target-headers.hmap -I/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/FMDB-all-non-framework-target-headers.hmap -ivfsoverlay /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/all-product-headers.yaml -iquote /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/FMDB-project-headers.hmap -I/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Products/Release-iphoneos/include -I/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/DerivedSources/armv7 -I/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/DerivedSources -F/Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Products/Release-iphoneos -MMD -MT dependencies -MF /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/Objects-normal/armv7/FMDatabase.d --serialize-diagnostics /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/Objects-normal/armv7/FMDatabase.dia -c /Users/MASKED_USER/fmdb27compile/Carthage/Checkouts/fmdb/src/fmdb/FMDatabase.m -o /Users/MASKED_USER/Library/Caches/org.carthage.CarthageKit/DerivedData/fmdb/v2.7/Build/Intermediates/fmdb.build/Release-iphoneos/FMDB\ iOS.build/Objects-normal/armv7/FMDatabase.o
/Users/MASKED_USER/fmdb27compile/Carthage/Checkouts/fmdb/src/fmdb/FMDatabase.m:1495:12: error: implicit conversion loses integer precision: 'sqlite3_int64' (aka 'long long') to 'long' [-Werror,-Wshorten-64-to-32]
    return sqlite3_value_int64(value);
    ~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

## Environment

* macOS 10.12.4 (16E195)
* Xcode 8.3.2 (8E2002)
* Carthage 0.23.0

**Comment by robertmryan** [comment]
@ccgus - I was able to consistently reproduce this warning (which was critical in my testing of the fix) by creating an iOS app, selecting an iPhone 5 simulator device. Then, because that's a 32-bit target, I get the warning consistently. I believe I've fixed this in 2.7.1, but take a look before we merge it.

**Comment by Kemenger** [comment]
YES，it will tip  this warning  in the 32-bit device(I test in my iPhone4S).I close the issue because my 
app only runs in 64-bit device.

**Comment by robertmryan** [comment]
It should be solved, regardless now, in [version 2.7.1](https://github.com/ccgus/fmdb/releases/tag/2.7.1).
